### PR TITLE
Use Varnish in a monorepo

### PR DESCRIPTION
**Pull request by Kirbeerus**
## Describe your PR

_Summarize your changes here : Some user may decide to use a monorepo, that means that they may have multiple applications inside one repository.  
In some case, they may want to use Varnish only for some applications, and not all of them.    
The current implementation of Varnish where you need to have a `/clevercloud/varnish.vcl` file at the root cause problem because Varnish would automatically be used with all the app.  
There is a way to resolve that problem by not putting the `varnish.vcl` file at the the root and creating a symlink using deployments hooks.

## Checklist

- [ ] My PR is related to an opened issue : #
- [x] I've read the [contributing guidelines](/CleverCloud/documentation/blob/main/CONTRIBUTING.md)

## Reviewers
_Who should review these changes?_ @

### CODE REVIEW

File: content/doc/administrate/cache.md [hunk 1]
```diff
@@ -57,3 +57,19 @@ If you already have a configuration for an older version of varnish, you can rea
...
 rename the file to `varnish.vcl` and move it in the `/clevercloud` folder.
+
+## Varnish with a monorepo
+
+If you use a monorepo, you may want to use Varnish for only some of the applications inside it.  
+If you have a `/clevercloud/varnish.vcl` file at the root of your monorepo, all of your applications automatically start using Varnish.
+
+To resolve this issue, you can create a symlink during the deployments.  
+First, you need to put your `varnish.vcl` file anywhere but at the root of your monorepo.  
+Then, you need to create a symlink inside a `CC_PRE_RUN_HOOK`.
```
Comment (Bencaddyro) [inline content/doc/administrate/cache.md:68]:
```suggestion
Then, you need to create a symlink inside a `CC_PRE_BUILD_HOOK`.  
```

File: content/doc/administrate/cache.md [hunk 2]
```diff
@@ -57,3 +57,19 @@ If you already have a configuration for an older version of varnish, you can rea
...
+If you use a monorepo, you may want to use Varnish for only some of the applications inside it.  
+If you have a `/clevercloud/varnish.vcl` file at the root of your monorepo, all of your applications automatically start using Varnish.
+
+To resolve this issue, you can create a symlink during the deployments.  
+First, you need to put your `varnish.vcl` file anywhere but at the root of your monorepo.  
+Then, you need to create a symlink inside a `CC_PRE_RUN_HOOK`.  
+Here is an example :
+
+```bash
+CC_PRE_RUN_HOOK="mkdir $APP_HOME/clevercloud; ln -s $APP_HOME/path/to/your/file/varnish.vcl $APP_HOME/clevercloud/varnish.vcl"
```
Comment (Bencaddyro) [inline content/doc/administrate/cache.md:72]:
```suggestion
CC_PRE_BUILD_HOOK="mkdir $APP_HOME/clevercloud; ln -s $APP_HOME/path/to/your/file/varnish.vcl $APP_HOME/clevercloud/varnish.vcl"
```

File: content/doc/administrate/cache.md [hunk 3]
```diff
@@ -57,3 +57,19 @@ If you already have a configuration for an older version of varnish, you can rea
...
+```
+
+This way, you only need to add the `CC_PRE_RUN_HOOK` when you want your app to use Varnish. If you don't, the application won't use it.
```
Comment (Bencaddyro) [inline content/doc/administrate/cache.md:75]:
```suggestion
This way, you only need to add the `CC_PRE_BUILD_HOOK` when you want your app to use Varnish. If you don't, the application won't use it.
```

File: content/doc/administrate/cache.md (hunk 1 above)
Comment (Bencaddyro) [reply content/doc/administrate/cache.md:68]: I don't think of any reason to do it on `PRE_RUN` so best practice to do it on `PRE_BUILD` instead

Is there any reason you put it on `PRE_RUN` ?

**Review by Bencaddyro — CHANGES_REQUESTED** [review]

File: content/doc/administrate/cache.md [hunk 4]
```diff
@@ -57,3 +57,19 @@ If you already have a configuration for an older version of varnish, you can rea
...
 use for your application. Create a `/clevercloud` folder at the root of your application if it does not exist,
 rename the file to `varnish.vcl` and move it in the `/clevercloud` folder.
+
+## Varnish with a monorepo
+
+If you use a monorepo, you may want to use Varnish for only some of the applications inside it.  
+If you have a `/clevercloud/varnish.vcl` file at the root of your monorepo, all of your applications automatically start using Varnish.
+
+To resolve this issue, you can create a symlink during the deployments.  
+First, you need to put your `varnish.vcl` file anywhere but at the root of your monorepo.
```
Comment (juliamrch) [inline content/doc/administrate/cache.md:67]:
```suggestion

1. Put your `varnish.vcl` file anywhere but at the root of your monorepo.  
```

File: content/doc/administrate/cache.md [hunk 5]
```diff
@@ -57,3 +57,19 @@ If you already have a configuration for an older version of varnish, you can rea
...
 rename the file to `varnish.vcl` and move it in the `/clevercloud` folder.
+
+## Varnish with a monorepo
+
+If you use a monorepo, you may want to use Varnish for only some of the applications inside it.  
+If you have a `/clevercloud/varnish.vcl` file at the root of your monorepo, all of your applications automatically start using Varnish.
+
+To resolve this issue, you can create a symlink during the deployments.  
+First, you need to put your `varnish.vcl` file anywhere but at the root of your monorepo.  
+Then, you need to create a symlink inside a `CC_PRE_BUILD_HOOK`.
```
Comment (juliamrch) [inline content/doc/administrate/cache.md:68]:
```suggestion
2. Create a symlink inside a `CC_PRE_BUILD_HOOK`.
  
```

File: content/doc/administrate/cache.md [hunk 6]
```diff
@@ -57,3 +57,19 @@ If you already have a configuration for an older version of varnish, you can rea
...
+To resolve this issue, you can create a symlink during the deployments.  
+First, you need to put your `varnish.vcl` file anywhere but at the root of your monorepo.  
+Then, you need to create a symlink inside a `CC_PRE_BUILD_HOOK`.  
+Here is an example :
+
+```bash
+CC_PRE_BUILD_HOOK="mkdir $APP_HOME/clevercloud; ln -s $APP_HOME/path/to/your/file/varnish.vcl $APP_HOME/clevercloud/varnish.vcl"
+```
+
+This way, you only need to add the `CC_PRE_BUILD_HOOK` when you want your app to use Varnish. If you don't, the application won't use it.
```
Comment (juliamrch) [inline content/doc/administrate/cache.md:75]:
```suggestion
Then add the `CC_PRE_BUILD_HOOK` in the variables of the app that needs to use Varnish. If you don't add this variable, the application won't use it.
```

It doesn't seem clear to me how you would handle the monorepo deployment. Many readers might think that the whole monoprepo containing several apps will run on a single runtime and might not understand  the variable instruction.

It looks like you are thinking about deploying a monorepo divided in several runtimes, using the variable `APP_FOLDER`, it that right?

**Review by juliamrch — COMMENTED** [review]
This is an interesting approach, but I'm not sure of the clarity or the relevance of this part of the doc. Added some comments, lmk what you think.

**Review by cnivolle — APPROVED** [review]

File: content/doc/administrate/cache.md (hunk 6 above)
Comment (Kirbeerus) [reply content/doc/administrate/cache.md]: This is indeed the case where someone has multiple app, sometime on different runtime and use `APP_FOLDER` to decide which app should be used.
For exemple, if a user has three different app in his monorepo, but only one need to use varnish, he needs to select that app with `APP_FOLDER` and follow the instructions in this PR.

If there was a variables like `CC_ENABLE_VARNISH=true`, the problem would be easily resolved. But right now, there is not a lot of way around that problem.

**Comment by juliamrch** [comment]
closing and reopening to recreate the review app (tests)

**Pull request merged**